barretenberg/cpp/src/barretenberg/vm2/simulation/context.cpp

```diff
@@ -0,0 +1,204 @@
+
+        // Tree State
+        // TreeSnapshots tree_state;
+    };
+};
+
+ContextStackEvent BaseContext::snapshot_context()
+{
+    auto [cd_offset, cd_size_offset] = this->get_calldata_info();
+    return ContextStackEvent{
+        .id = context_id,
+        .parent_id = this->get_parent_id(),
+
+        // State
+        .pc = next_pc, // NOTE: We need the next pc here so it should be set correctly
+        .msg_sender = msg_sender,
+        .contract_addr = address,
+        .is_static = is_static,
+
+        // Calldata info
+        .cd_addr = cd_offset,
+        .cd_size_addr = cd_size_offset,
+
+        // Gas
+        // .l2_gas_used;  -- this is also probably going to be "next gas used" or "gas_used_after"
+        // .l2_gas_limit;
+        // .da_gas_used;
+        // .da_gas_limit;
+
+        // Tree State
+        // TreeSnapshots tree_state;
+    };
+}
+
+// This pushes a new context event to the event emitter.
+// And sets the last_context_event_ptr to the last event in the events vector.
+// If this function is called repeatedly without a separate emit_current_context call, it will be a no-op
+void BaseContext::reserve_context_event()
+{
+    emit_current_context();
+    last_context_event = &events.last();
+}
+
+// Used to emit a new context event to the event emitter.
+// If we have a last_context_event_ptr cached, no event is emitted, bu the last_context_event_ptr is cleared
+void BaseContext::emit_current_context()
+{
+    if (last_context_event == nullptr) {
+        auto event = serialize_context();
+        events.emit(std::move(event));
+    } else {
+        last_context_event = nullptr;
+    }
+}
+
+void BaseContext::emit_ctx_stack_event()
+{
+    auto event = snapshot_context();
+    ctx_stack_events.emit(std::move(event));
+}
+
+/**************************************************************************************************
+ *                                   EnqueuedCallContext Class
+ **************************************************************************************************/
+std::vector<FF> EnqueuedCallContext::get_calldata(uint32_t cd_offset, uint32_t size)
+{
+    return { calldata.begin() + cd_offset, calldata.begin() + cd_offset + size };
+}
+
+std::vector<FF> EnqueuedCallContext::get_returndata(uint32_t rd_offset, uint32_t size)
+{
+    return { top_level_returndata.begin() + rd_offset, top_level_returndata.begin() + rd_offset + size };
+}
+
+std::pair<uint32_t, uint32_t> EnqueuedCallContext::get_calldata_info()
+{
+    return { 0, 0 };
+}
+
+std::pair<uint32_t, uint32_t> EnqueuedCallContext::get_returndata_info()
+{
+    return { 0, 0 };
+}
+
+/**************************************************************************************************
+ *                                   NestedContext Class
+ **************************************************************************************************/
+std::vector<FF> NestedContext::get_calldata(uint32_t cd_offset, uint32_t size)
```

Comment:
These need to be revisited but are left here to show how we will interact with the parent/child contexts